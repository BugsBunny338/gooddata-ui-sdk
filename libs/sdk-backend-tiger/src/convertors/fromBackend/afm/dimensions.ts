// (C) 2019-2020 GoodData Corporation
import {
    IAttributeDescriptor,
    IMeasureGroupDescriptor,
    IMeasureDescriptor,
    IDimensionDescriptor,
} from "@gooddata/sdk-backend-spi";
import { Execution } from "@gooddata/api-client-tiger";

import isAttributeHeader = Execution.isAttributeHeader;

const DEFAULT_FORMAT = "#,#.##";

function transformDimension(dim: Execution.IResultDimension): IDimensionDescriptor {
    return {
        headers: dim.headers.map((header, headerIdx) => {
            const h = header;

            if (isAttributeHeader(h)) {
                /*
                 * Funny stuff #1: we have to set 'uri' to some made-up value resembling the URIs sent by bear. This
                 *  is because pivot table relies on the format of URIs. Ideally we would refactor pivot table to
                 *  not care about this however this aspect is like a couple of eggs that hold the pivot spaghetti
                 *  together - cannot be easily untangled.
                 */
                const attrDescriptor: IAttributeDescriptor = {
                    attributeHeader: {
                        uri: `/obj/${headerIdx}`,
                        identifier: h.attributeHeader.identifier,
                        formOf: {
                            identifier: h.attributeHeader.formOf.identifier,
                            name: h.attributeHeader.formOf.name,
                            uri: `/obj/${headerIdx}`,
                        },
                        localIdentifier: h.attributeHeader.localIdentifier,
                        name: h.attributeHeader.name,
                    },
                };

                return attrDescriptor;
            } else {
                /*
                 * Funny stuff #2: tiger does not send name & format according to the contract (which is inspired
                 *  by bear behavior). The code must reconciliate as follows:
                 *
                 *  -  if name does not come from tiger, then default the name to localIdentifier
                 *  -  if format does not come from tiger, then default to a hardcoded format
                 */
                const measureDescriptor: IMeasureGroupDescriptor = {
                    measureGroupHeader: {
                        items: h.measureGroupHeader.items.map((m) => {
                            const newItem: IMeasureDescriptor = {
                                measureHeaderItem: {
                                    localIdentifier: m.measureHeaderItem.localIdentifier,
                                    name: m.measureHeaderItem.name ?? m.measureHeaderItem.localIdentifier,
                                    format: m.measureHeaderItem.format ?? DEFAULT_FORMAT,
                                },
                            };

                            return newItem;
                        }),
                    },
                };

                return measureDescriptor;
            }
        }),
    };
}

/**
 * Transforms dimensions in the result provided by backend to the unified model used in SDK.
 *
 * @param dimensions - dimensions from execution result
 * @returns dimensions as used in the unified model
 */
export function transformResultDimensions(dimensions: Execution.IResultDimension[]): IDimensionDescriptor[] {
    return dimensions.map(transformDimension);
}
