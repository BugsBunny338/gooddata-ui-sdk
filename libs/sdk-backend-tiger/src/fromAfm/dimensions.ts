// (C) 2019 GoodData Corporation
import {
    IAttributeHeader,
    IMeasureGroupHeader,
    IMeasureHeaderItem,
    IResultDimension,
} from "@gooddata/sdk-backend-spi";
import { Execution } from "../gd-tiger-model/Execution";
import isAttributeHeader = Execution.isAttributeHeader;

function transformDimension(dim: Execution.IResultDimension): IResultDimension {
    return {
        headers: dim.headers.map(header => {
            const h = header;

            if (isAttributeHeader(h)) {
                // TODO: SDK8: make our mind about whether URIs should be optional in attribute headers
                //  perhaps it makes sense. perhaps the URIs should be always present and should be coming from
                //  new stack as well? anyway, making URIs optional in the domain model will mean waterfall of fun
                //  changes across public interface of the SDK (drilling etc)
                const attrHeader: IAttributeHeader = {
                    attributeHeader: {
                        uri: `/fakeAttrUri/${h.attributeHeader.identifier}`,
                        identifier: h.attributeHeader.identifier,
                        formOf: {
                            identifier: h.attributeHeader.identifier,
                            name: h.attributeHeader.identifier,
                            uri: `/fakeDfUri/${h.attributeHeader.identifier}`,
                        },
                        localIdentifier: h.attributeHeader.identifier,
                        name: h.attributeHeader.identifier,
                    },
                };

                return attrHeader;
            } else {
                const measureHeader: IMeasureGroupHeader = {
                    measureGroupHeader: {
                        items: h.measureGroupHeader.items.map(m => {
                            const newItem: IMeasureHeaderItem = {
                                measureHeaderItem: {
                                    localIdentifier: m.measureHeaderItem.localIdentifier,
                                    name: m.measureHeaderItem.name,
                                    format: m.measureHeaderItem.format,
                                },
                            };

                            return newItem;
                        }),
                    },
                };

                return measureHeader;
            }
        }),
    };
}

/**
 * Transforms dimensions in the result provided by backend to the unified model used in SDK.
 *
 * @param dimensions - dimensions from execution result
 * @returns dimensions as used in the unified model
 */
export function transformResultDimensions(dimensions: Execution.IResultDimension[]): IResultDimension[] {
    return dimensions.map(transformDimension);
}
