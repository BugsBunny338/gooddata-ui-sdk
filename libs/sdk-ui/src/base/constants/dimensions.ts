// (C) 2007-2018 GoodData Corporation
export const MEASUREGROUP = "measureGroup";

export const VIEW_BY_DIMENSION_INDEX = 1;
export const STACK_BY_DIMENSION_INDEX = 0;
export const PARENT_ATTRIBUTE_INDEX = 0; // first attribute in 'View By' panel
export const PRIMARY_ATTRIBUTE_INDEX = 1; // second attribute in 'View By' panel
