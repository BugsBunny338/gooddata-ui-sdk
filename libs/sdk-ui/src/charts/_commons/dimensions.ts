// (C) 2019 GoodData Corporation

import {
    attributeLocalId,
    bucketAttribute,
    bucketAttributes,
    bucketIsEmpty,
    bucketsAttributes,
    bucketsFind,
    IBucket,
    IDimension,
    IExecutionDefinition,
    MeasureGroupIdentifier,
    newTwoDimensional,
} from "@gooddata/sdk-model";
import { BucketNames } from "../../base";

function isStackedChart(buckets: IBucket[]) {
    return !bucketIsEmpty(bucketsFind(buckets, BucketNames.STACK));
}

function stackedDimensions(buckets: IBucket[]): IDimension[] {
    const viewBucket = bucketsFind(buckets, BucketNames.ATTRIBUTE);
    const stackBucket = bucketsFind(buckets, BucketNames.STACK);

    const viewByAttributes = bucketAttributes(viewBucket);
    const stackByAttribute = bucketAttribute(stackBucket);

    const stackByAttributeLocalIdentifier = stackByAttribute
        ? stackByAttribute.attribute.localIdentifier
        : undefined;

    const viewByAttributeLocalIdentifiers = viewByAttributes && viewByAttributes.map(attributeLocalId);

    return newTwoDimensional(
        stackByAttributeLocalIdentifier ? [stackByAttributeLocalIdentifier] : [],
        viewByAttributeLocalIdentifiers
            ? [...viewByAttributeLocalIdentifiers, MeasureGroupIdentifier]
            : [MeasureGroupIdentifier],
    );
}

export function defaultDimensions(def: IExecutionDefinition): IDimension[] {
    return newTwoDimensional([MeasureGroupIdentifier], bucketsAttributes(def.buckets).map(attributeLocalId));
}

export function stackedChartDimensions(def: IExecutionDefinition): IDimension[] {
    const { buckets } = def;
    return isStackedChart(buckets) ? stackedDimensions(buckets) : defaultDimensions(def);
}

export function pointyChartDimensions(def: IExecutionDefinition): IDimension[] {
    return newTwoDimensional(bucketsAttributes(def.buckets).map(attributeLocalId), [MeasureGroupIdentifier]);
}

export function roundChartDimensions(def: IExecutionDefinition): IDimension[] {
    const attributes = bucketsAttributes(def.buckets).map(attributeLocalId);

    if (attributes.length === 0) {
        return newTwoDimensional([], [MeasureGroupIdentifier]);
    }

    return newTwoDimensional([MeasureGroupIdentifier], attributes);
}

export function heatmapDimensions(def: IExecutionDefinition): IDimension[] {
    const view: IBucket = bucketsFind(def.buckets, BucketNames.VIEW);
    const stack: IBucket = bucketsFind(def.buckets, BucketNames.STACK);

    if (bucketIsEmpty(stack)) {
        return newTwoDimensional(bucketAttributes(view).map(attributeLocalId), [MeasureGroupIdentifier]);
    }

    return newTwoDimensional(
        bucketAttributes(view).map(attributeLocalId),
        bucketAttributes(stack)
            .map(attributeLocalId)
            .concat([MeasureGroupIdentifier]),
    );
}

export function treemapDimensions(def: IExecutionDefinition): IDimension[] {
    const attributes = bucketsAttributes(def.buckets);

    if (attributes.length === 1) {
        return newTwoDimensional([MeasureGroupIdentifier], attributes.map(attributeLocalId));
    }

    return newTwoDimensional(attributes.map(attributeLocalId), [MeasureGroupIdentifier]);
}
