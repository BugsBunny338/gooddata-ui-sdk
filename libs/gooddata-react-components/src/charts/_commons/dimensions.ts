// (C) 2019 GoodData Corporation

import {
    IBucket,
    IDimension,
    IAttribute,
    attributeId,
    bucketsAttributes,
    bucketAttributes,
} from "@gooddata/sdk-model";
import { ATTRIBUTE, STACK, VIEW } from "../../constants/bucketNames";
import { MEASUREGROUP } from "../../constants/dimensions";

function isStackedChart(buckets: IBucket[]) {
    return buckets.some(bucket => bucket.localIdentifier === STACK && bucket.items.length > 0);
}

function stackedDimensions(buckets: IBucket[]): IDimension[] {
    const viewBucket = buckets.find(bucket => bucket.localIdentifier === ATTRIBUTE);
    const stackBucket = buckets.find(bucket => bucket.localIdentifier === STACK);

    const viewByAttributes = viewBucket && (viewBucket.items as IAttribute[]);
    const stackByAttribute = stackBucket && (stackBucket.items[0] as IAttribute);

    const stackByAttributeLocalIdentifier = stackByAttribute
        ? stackByAttribute.attribute.localIdentifier
        : undefined;

    const viewByAttributeLocalIdentifiers = viewByAttributes && viewByAttributes.map(attributeId);

    return [
        {
            itemIdentifiers: stackByAttributeLocalIdentifier ? [stackByAttributeLocalIdentifier] : [],
        },
        {
            itemIdentifiers: viewByAttributeLocalIdentifiers
                ? [...viewByAttributeLocalIdentifiers, MEASUREGROUP]
                : [MEASUREGROUP],
        },
    ];
}

export function defaultDimensions(buckets: IBucket[]): IDimension[] {
    return [
        {
            itemIdentifiers: [MEASUREGROUP],
        },
        {
            itemIdentifiers: bucketsAttributes(buckets).map(attributeId),
        },
    ];
}

export function stackedChartDimensions(buckets: IBucket[]): IDimension[] {
    return isStackedChart(buckets) ? stackedDimensions(buckets) : defaultDimensions(buckets);
}

export function pointyChartDimensions(buckets: IBucket[]): IDimension[] {
    return [
        {
            itemIdentifiers: bucketsAttributes(buckets).map(attributeId),
        },
        {
            itemIdentifiers: [MEASUREGROUP],
        },
    ];
}

export function roundChartDimensions(buckets: IBucket[]): IDimension[] {
    const attributes = bucketsAttributes(buckets).map(attributeId);

    if (attributes.length === 0) {
        return [
            {
                itemIdentifiers: [],
            },
            {
                itemIdentifiers: [MEASUREGROUP],
            },
        ];
    }

    return [
        {
            itemIdentifiers: [MEASUREGROUP],
        },
        {
            itemIdentifiers: attributes,
        },
    ];
}

export function heatmapDimensions(buckets: IBucket[]): IDimension[] {
    const view: IBucket = buckets.find(bucket => bucket.localIdentifier === VIEW);
    const stack: IBucket = buckets.find(bucket => bucket.localIdentifier === STACK);

    const hasNoStacks = !stack || !stack.items || stack.items.length === 0;

    if (hasNoStacks) {
        return [
            {
                itemIdentifiers: bucketAttributes(view).map(attributeId),
            },
            {
                itemIdentifiers: [MEASUREGROUP],
            },
        ];
    }

    return [
        {
            itemIdentifiers: bucketAttributes(view).map(attributeId),
        },
        {
            itemIdentifiers: bucketAttributes(stack)
                .map(attributeId)
                .concat([MEASUREGROUP]),
        },
    ];
}
