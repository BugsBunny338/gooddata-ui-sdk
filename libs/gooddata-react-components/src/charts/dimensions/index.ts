// (C) 2019 GoodData Corporation

import { IBucket, IDimension, IAttribute, attributeId, bucketsAttributes } from "@gooddata/sdk-model";
import { ATTRIBUTE, STACK } from "../../constants/bucketNames";
import { MEASUREGROUP } from "../../constants/dimensions";

function isStackedChart(buckets: IBucket[]) {
    return buckets.some(bucket => bucket.localIdentifier === STACK && bucket.items.length > 0);
}

function stackedDimensions(buckets: IBucket[]): IDimension[] {
    const viewBucket = buckets.find(bucket => bucket.localIdentifier === ATTRIBUTE);
    const stackBucket = buckets.find(bucket => bucket.localIdentifier === STACK);

    const viewByAttributes = viewBucket && (viewBucket.items as IAttribute[]);
    const stackByAttribute = stackBucket && (stackBucket.items[0] as IAttribute);

    const stackByAttributeLocalIdentifier = stackByAttribute
        ? stackByAttribute.attribute.localIdentifier
        : undefined;

    const viewByAttributeLocalIdentifiers = viewByAttributes && viewByAttributes.map(attributeId);

    return [
        {
            itemIdentifiers: stackByAttributeLocalIdentifier ? [stackByAttributeLocalIdentifier] : [],
        },
        {
            itemIdentifiers: viewByAttributeLocalIdentifiers
                ? [...viewByAttributeLocalIdentifiers, MEASUREGROUP]
                : [MEASUREGROUP],
        },
    ];
}

export function defaultDimensions(buckets: IBucket[]): IDimension[] {
    return [
        {
            itemIdentifiers: [MEASUREGROUP],
        },
        {
            itemIdentifiers: bucketsAttributes(buckets).map(attributeId),
        },
    ];
}

export function stackedChartDimensions(buckets: IBucket[]): IDimension[] {
    return isStackedChart(buckets) ? stackedDimensions(buckets) : defaultDimensions(buckets);
}
