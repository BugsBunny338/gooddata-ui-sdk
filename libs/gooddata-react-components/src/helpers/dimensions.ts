// (C) 2007-2019 GoodData Corporation
import get = require("lodash/get");
import { AFM, VisualizationObject } from "@gooddata/typings";
import { VisType, VisualizationTypes } from "../constants/visualizationTypes";
import { MEASUREGROUP } from "../constants/dimensions";
import { ATTRIBUTE, COLUMNS, MEASURES, SEGMENT, STACK, TREND, VIEW } from "../constants/bucketNames";
import { VIEW_BY_ATTRIBUTES_LIMIT } from "../components/visualizations/chart/constants";
import {
    attributeId,
    bucketAttribute,
    bucketAttributes,
    bucketIsEmpty,
    IDimension,
    IInsight,
    insightAttributes,
    insightBucket,
    insightMeasures,
    insightTotals,
    ITotal,
} from "@gooddata/sdk-model";

export function getDimensionTotals(bucket: VisualizationObject.IBucket): ITotal[] {
    const bucketTotals: VisualizationObject.IVisualizationTotal[] = get<
        VisualizationObject.IBucket,
        "totals",
        VisualizationObject.IVisualizationTotal[]
    >(bucket, "totals", []);
    return bucketTotals.map(
        (total: VisualizationObject.IVisualizationTotal): AFM.ITotalItem => {
            return {
                measureIdentifier: total.measureIdentifier,
                type: total.type,
                attributeIdentifier: total.attributeIdentifier,
            };
        },
    );
}

export function getPivotTableDimensions(insight: IInsight): IDimension[] {
    const row = insightBucket(insight, ATTRIBUTE);
    const columns = insightBucket(insight, COLUMNS);
    const measures = insightBucket(insight, MEASURES);

    const rowAttributeIds = bucketAttributes(row).map(attributeId);
    const columnAttributeIds = bucketAttributes(columns).map(attributeId);

    const measuresItemIdentifiers = !bucketIsEmpty(measures) ? [MEASUREGROUP] : [];

    const totals = insightTotals(insight);
    const totalsProp = totals.length ? { totals } : {};

    return [
        {
            itemIdentifiers: rowAttributeIds,
            ...totalsProp,
        },
        {
            itemIdentifiers: [...columnAttributeIds, ...measuresItemIdentifiers],
        },
    ];
}

function getPieOrDonutDimensions(insight: IInsight): IDimension[] {
    const viewBy = insightBucket(insight, VIEW);

    if (!bucketIsEmpty(viewBy)) {
        return [
            {
                itemIdentifiers: [MEASUREGROUP],
            },
            {
                itemIdentifiers: bucketAttributes(viewBy).map(attributeId),
            },
        ];
    }

    return [
        {
            itemIdentifiers: [],
        },
        {
            itemIdentifiers: [MEASUREGROUP],
        },
    ];
}

function getBarDimensions(insight: IInsight): IDimension[] {
    const viewBy = insightBucket(insight, VIEW);
    const stack = insightBucket(insight, STACK);

    if (bucketIsEmpty(stack)) {
        return [
            {
                itemIdentifiers: [MEASUREGROUP],
            },
            {
                itemIdentifiers: bucketAttributes(viewBy).map(attributeId),
            },
        ];
    }

    return [
        {
            itemIdentifiers: bucketAttributes(stack).map(attributeId),
        },
        {
            itemIdentifiers: [...bucketAttributes(viewBy).map(attributeId), MEASUREGROUP],
        },
    ];
}

function getAreaDimensions(insight: IInsight): IDimension[] {
    const viewByAttributes = bucketAttributes(insightBucket(insight, VIEW));
    const stack = insightBucket(insight, STACK);

    if (viewByAttributes.length > 1) {
        // only take first two view items
        const [viewItemIdentifier, stackItemIdentifier] = viewByAttributes
            .slice(0, VIEW_BY_ATTRIBUTES_LIMIT)
            .map(attributeId);
        return [
            {
                itemIdentifiers: [stackItemIdentifier],
            },
            {
                itemIdentifiers: [viewItemIdentifier, MEASUREGROUP],
            },
        ];
    }

    if (bucketIsEmpty(stack)) {
        return [
            {
                itemIdentifiers: [MEASUREGROUP],
            },
            {
                itemIdentifiers: viewByAttributes.map(attributeId),
            },
        ];
    }

    return [
        {
            itemIdentifiers: bucketAttributes(stack).map(attributeId),
        },
        {
            itemIdentifiers: [...viewByAttributes.map(attributeId), MEASUREGROUP],
        },
    ];
}

function getLineDimensions(insight: IInsight): IDimension[] {
    const trend = insightBucket(insight, TREND);
    const segment = insightBucket(insight, SEGMENT);

    if (bucketIsEmpty(segment)) {
        return [
            {
                itemIdentifiers: [MEASUREGROUP],
            },
            {
                itemIdentifiers: bucketAttributes(trend).map(attributeId),
            },
        ];
    }

    return [
        {
            itemIdentifiers: bucketAttributes(segment).map(attributeId),
        },
        {
            itemIdentifiers: [...bucketAttributes(trend).map(attributeId), MEASUREGROUP],
        },
    ];
}

export function getHeadlinesDimensions(): IDimension[] {
    return [{ itemIdentifiers: ["measureGroup"] }];
}

function getScatterDimensions(insight: IInsight): IDimension[] {
    const attribute = insightBucket(insight, ATTRIBUTE);

    if (!bucketIsEmpty(attribute)) {
        return [
            {
                itemIdentifiers: bucketAttributes(attribute).map(attributeId),
            },
            {
                itemIdentifiers: [MEASUREGROUP],
            },
        ];
    }

    return [
        {
            itemIdentifiers: [],
        },
        {
            itemIdentifiers: [MEASUREGROUP],
        },
    ];
}

// Heatmap
export function getHeatmapDimensionsFromMdObj(insight: IInsight): IDimension[] {
    return getHeatmapDimensionsFromBuckets(insight);
}

export function getHeatmapDimensionsFromBuckets(insight: IInsight): IDimension[] {
    const view = insightBucket(insight, VIEW);
    const stack = insightBucket(insight, STACK);

    if (bucketIsEmpty(stack)) {
        return [
            {
                itemIdentifiers: bucketAttributes(view).map(attributeId),
            },
            {
                itemIdentifiers: [MEASUREGROUP],
            },
        ];
    }

    return [
        {
            itemIdentifiers: bucketAttributes(view).map(attributeId),
        },
        {
            itemIdentifiers: [...bucketAttributes(stack).map(attributeId), MEASUREGROUP],
        },
    ];
}

function getBubbleDimensions(insight: IInsight): IDimension[] {
    const view = insightBucket(insight, VIEW);
    const stack = insightBucket(insight, STACK);

    if (bucketIsEmpty(stack)) {
        return [
            {
                itemIdentifiers: bucketAttributes(view).map(attributeId),
            },
            {
                itemIdentifiers: [MEASUREGROUP],
            },
        ];
    }

    return [
        {
            itemIdentifiers: [
                ...bucketAttributes(view).map(attributeId),
                ...bucketAttributes(stack).map(attributeId),
            ],
        },
        {
            itemIdentifiers: [MEASUREGROUP],
        },
    ];
}

/**
 * generateDimensions
 * is a function that generates dimensions based on buckets and visualization objects.
 * WARNING: It duplicates logic from pluggable visualizations.
 *          Remove once react components support pluggable visualizations.
 * @param insight - insight being visualized
 * @param type:VisType - visualization type string
 * @internal
 */
export function generateDimensions(insight: IInsight, type: VisType): IDimension[] {
    switch (type) {
        case VisualizationTypes.TABLE: {
            return getPivotTableDimensions(insight);
        }
        case VisualizationTypes.PIE:
        case VisualizationTypes.DONUT:
        case VisualizationTypes.FUNNEL: {
            return getPieOrDonutDimensions(insight);
        }
        case VisualizationTypes.TREEMAP: {
            return getTreemapDimensionsFromMdObj(insight);
        }

        case VisualizationTypes.LINE: {
            return getLineDimensions(insight);
        }

        case VisualizationTypes.AREA: {
            return getAreaDimensions(insight);
        }

        case VisualizationTypes.BAR:
        case VisualizationTypes.COMBO:
        case VisualizationTypes.COMBO2:
        case VisualizationTypes.COLUMN: {
            return getBarDimensions(insight);
        }
        case VisualizationTypes.HEADLINE: {
            return getHeadlinesDimensions();
        }
        case VisualizationTypes.SCATTER: {
            return getScatterDimensions(insight);
        }
        case VisualizationTypes.HEATMAP: {
            return getHeatmapDimensionsFromMdObj(insight);
        }
        case VisualizationTypes.BUBBLE: {
            return getBubbleDimensions(insight);
        }
    }
    return [];
}

export function generateStackedDimensions(insight: IInsight): IDimension[] {
    const viewAttributes = bucketAttributes(insightBucket(insight, ATTRIBUTE));
    const stackAttribute = bucketAttribute(insightBucket(insight, STACK));

    return [
        {
            itemIdentifiers: stackAttribute ? [attributeId(stackAttribute)] : [],
        },
        {
            itemIdentifiers: [...viewAttributes.map(attributeId), MEASUREGROUP],
        },
    ];
}

// for LineChart, AreaChart, BarChart, ColumnChart
export function generateDefaultDimensions(insight: IInsight): IDimension[] {
    return [
        {
            itemIdentifiers: [MEASUREGROUP],
        },
        {
            itemIdentifiers: insightAttributes(insight).map(attributeId),
        },
    ];
}

export function isStackedChart(buckets: VisualizationObject.IBucket[], stackedBuckedName: string = STACK) {
    return buckets.some(bucket => bucket.localIdentifier === stackedBuckedName && bucket.items.length > 0);
}

// for ScatterPlot and BubbleChart
export function generateDefaultDimensionsForPointsCharts(insight: IInsight): IDimension[] {
    return [
        {
            itemIdentifiers: insightAttributes(insight).map(attributeId),
        },
        {
            itemIdentifiers: [MEASUREGROUP],
        },
    ];
}

// for PieChart, DonutChart
export const generateDefaultDimensionsForRoundChart = (insight: IInsight): IDimension[] => {
    const attributes = insightAttributes(insight);

    if (attributes.length === 0) {
        return [
            {
                itemIdentifiers: [],
            },
            {
                itemIdentifiers: [MEASUREGROUP],
            },
        ];
    }

    return [
        {
            itemIdentifiers: [MEASUREGROUP],
        },
        {
            itemIdentifiers: attributes.map(attributeId),
        },
    ];
};

// Treemap
export function getTreemapDimensionsFromMdObj(insight: IInsight): IDimension[] {
    return getTreemapDimensionsFromBuckets(insight);
}

export function getTreemapDimensionsFromBuckets(insight: IInsight): IDimension[] {
    return getTreemapDimensionsFromAFM(insight);
}

export function getTreemapDimensionsFromAFM(insight: IInsight): IDimension[] {
    const attributes = insightAttributes(insight);

    if (attributes.length === 1) {
        return [
            {
                itemIdentifiers: [MEASUREGROUP],
            },
            {
                itemIdentifiers: attributes.map(attributeId),
            },
        ];
    }

    return [
        {
            itemIdentifiers: attributes.map(attributeId),
        },
        {
            itemIdentifiers: [MEASUREGROUP],
        },
    ];
}

export function getGeneralDimensionsFromAFM(insight: IInsight): IDimension[] {
    const attributes = insightAttributes(insight);
    const measures = insightMeasures(insight);
    const dimensions = [];

    if (attributes.length > 0) {
        dimensions.push({
            itemIdentifiers: attributes.map(attributeId),
        });
    }
    if (measures.length > 0) {
        dimensions.push({
            itemIdentifiers: [MEASUREGROUP],
        });
    }
    return dimensions;
}
