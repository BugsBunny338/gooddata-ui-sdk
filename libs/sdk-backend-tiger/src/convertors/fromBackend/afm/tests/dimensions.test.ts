// (C) 2020 GoodData Corporation

import { transformResultDimensions } from "../dimensions";
import { mockDimensions } from "./dimensions.fixture";
import { emptyDef, idRef, newDefForItems, newMeasure } from "@gooddata/sdk-model";

describe("transformResultDimensions", () => {
    it("should fill in uris and refs for attribute descriptors", () => {
        expect(transformResultDimensions(mockDimensions, emptyDef("test"))).toMatchSnapshot();
    });

    it("should fill in uris and refs for attribute descriptors and simple measure descriptors", () => {
        expect(
            transformResultDimensions(
                mockDimensions,
                newDefForItems("test", [
                    newMeasure(idRef("measureIdentifier", "measure"), (m) => m.localId("measureLocalId")),
                ]),
            ),
        ).toMatchSnapshot();
    });
});
