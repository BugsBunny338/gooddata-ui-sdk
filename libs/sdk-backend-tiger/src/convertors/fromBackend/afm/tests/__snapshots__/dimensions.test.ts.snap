// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformResultDimensions should fill in uris and refs for attribute descriptors 1`] = `
Array [
  Object {
    "headers": Array [
      Object {
        "measureGroupHeader": Object {
          "items": Array [
            Object {
              "measureHeaderItem": Object {
                "format": "#,##0.00",
                "identifier": undefined,
                "localIdentifier": "measureLocalId",
                "name": "measureLocalId",
                "ref": undefined,
                "uri": undefined,
              },
            },
          ],
        },
      },
    ],
  },
  Object {
    "headers": Array [
      Object {
        "attributeHeader": Object {
          "formOf": Object {
            "identifier": "event_date.quarter",
            "name": "event_date - Quarter/Year",
            "ref": Object {
              "identifier": "event_date.quarter",
              "type": "attribute",
            },
            "uri": "/obj/0",
          },
          "identifier": "event_date.quarter.label",
          "localIdentifier": "fd48e8fd32b54b1baba3e3ccdd719f26",
          "name": "event_date - Quarter/Year",
          "ref": Object {
            "identifier": "event_date.quarter.label",
            "type": "displayForm",
          },
          "uri": "/obj/0",
        },
      },
    ],
  },
]
`;

exports[`transformResultDimensions should fill in uris and refs for attribute descriptors and simple measure descriptors 1`] = `
Array [
  Object {
    "headers": Array [
      Object {
        "measureGroupHeader": Object {
          "items": Array [
            Object {
              "measureHeaderItem": Object {
                "format": "#,##0.00",
                "identifier": "measureIdentifier",
                "localIdentifier": "measureLocalId",
                "name": "measureLocalId",
                "ref": Object {
                  "identifier": "measureIdentifier",
                  "type": "measure",
                },
                "uri": "/obj/0",
              },
            },
          ],
        },
      },
    ],
  },
  Object {
    "headers": Array [
      Object {
        "attributeHeader": Object {
          "formOf": Object {
            "identifier": "event_date.quarter",
            "name": "event_date - Quarter/Year",
            "ref": Object {
              "identifier": "event_date.quarter",
              "type": "attribute",
            },
            "uri": "/obj/0",
          },
          "identifier": "event_date.quarter.label",
          "localIdentifier": "fd48e8fd32b54b1baba3e3ccdd719f26",
          "name": "event_date - Quarter/Year",
          "ref": Object {
            "identifier": "event_date.quarter.label",
            "type": "displayForm",
          },
          "uri": "/obj/0",
        },
      },
    ],
  },
]
`;
